### PR TITLE
Add TSAPI functions TSRemapFrom/ToUrlGet().  Allow TSMBuffer pointer passed to TSUrlStringGet() to be null.

### PR DESCRIPTION
**Pull request by ywkaras**

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
Overall looks reasonable, although see the API discussion in the mailing list.

Also, documentation for these API functions in `doc/developer-guide/api/functions` would be helpful.

**Comment by SolidWallOfCode** [comment]
I understand why you used pointer to member, but I'm not sure how popular that will be vs. copy and paste of the validation code.

**Comment by ywkaras** [comment]
There are so many ways to make code more stupid.  Best to get feedback on which particular sort of stupid the community finds most desirable.

**Comment by ywkaras** [comment]
Documentation and automated testing are needed.  I'm trying to avoid repeating the pattern of doing a lot of work on stuff that just ends up shit-canned.

File: proxy/InkAPI.cc [hunk 1]
```diff
@@ -2137,7 +2137,8 @@ TSUrlLengthGet(TSMBuffer bufp, TSMLoc obj)
 char *
 TSUrlStringGet(TSMBuffer bufp, TSMLoc obj, int *length)
 {
-  sdk_assert(sdk_sanity_check_mbuffer(bufp) == TS_SUCCESS);
+  // bufp is not actually used anymore, so it can be null.
+  sdk_assert((sdk_sanity_check_mbuffer(bufp) == TS_SUCCESS) or (sdk_sanity_check_null_ptr(bufp) == TS_SUCCESS));
```
Comment (SolidWallOfCode) [inline proxy/InkAPI.cc:2141]: I'd probably do the cheaper null check first.

Comment (SolidWallOfCode) [reply proxy/InkAPI.cc:2141]: I'm a bit confused - `sdk_sanity_check_null_ptr` returns `TS_SUCCESS` if the argument is *not* `nullptr`. Therefore I don't think this allows a `nullptr`.

**Review by SolidWallOfCode — CHANGES_REQUESTED** [review]
The nullptr checks seem conflicted.

File: proxy/InkAPI.cc (hunk 1 above)
Comment (ywkaras) [reply proxy/InkAPI.cc:2141]: You're looking for love in all the wrong places, review here:  https://github.com/apache/trafficserver/pull/3145

This PR is obsolete.

**Pull request closed without merging**